river/PointerMapping.zig

```diff
@@ -14,13 +14,72 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
+const Self = @This();
+
+const std = @import("std");
 const wlr = @import("wlroots");
 
-pub const Action = enum {
+const util = @import("util.zig");
+
+pub const ActionType = enum {
     move,
     resize,
+    command,
+};
+
+pub const Action = union(ActionType) {
+    move: void,
+    resize: void,
+    command: []const [:0]const u8,
 };
 
 event_code: u32,
 modifiers: wlr.Keyboard.ModifierMask,
 action: Action,
+
+pub fn init(
+    event_code: u32,
+    modifiers: wlr.Keyboard.ModifierMask,
+    action_type: ActionType,
+    command_args: []const [:0]const u8,
+) !Self {
+    const action: Action = switch (action_type) {
+        ActionType.move => Action.move,
+        ActionType.resize => Action.resize,
+        ActionType.command => blk: {
+            var owned_args: std.ArrayListUnmanaged([:0]const u8) = .{};
+            try owned_args.ensureTotalCapacity(util.gpa, command_args.len);
+
+            errdefer {
+                for (owned_args.items) |arg| util.gpa.free(arg);
+                owned_args.deinit(util.gpa);
+            }
+            for (command_args) |arg| {
+                const owned = try util.gpa.dupeZ(u8, arg);
+                owned_args.appendAssumeCapacity(owned);
+            }
+
+            const action: Action = Action{ .command = owned_args.toOwnedSlice(util.gpa) };
+
+            owned_args.deinit(util.gpa);
```

Comment:
I do not think this is correct. It seems that a double free may occur from the `errdefer` block.